Add JOSS paper (#804)
Materials-Consortia/optimade-python-tools: paper/paper.md

---

> title: '`optimade-python-tools`: a Python library for implementing and consuming materials data via OPTIMADE APIs'

Author: ```suggestion
title: '`optimade-python-tools`: a Python library for serving and consuming materials data via OPTIMADE APIs'
```

   index: 1
 - name: Theory of Condensed Matter Group, Cavendish Laboratory, University of Cambridge, J. J. Thomson Avenue, Cambridge, CB3 0HE, U.K.
   index: 2
 - name: EPFL
   index: 3
 - name: Lawrence Berkeley National Laboratory, Berkeley, CA, USA
   index: 4
 - name: FHI
   index: 5
 - name: Namur
   index: 6
 - name: Polyneme LLC
   index: 7
 - name: KCL
   index: 8

date: June 2021
bibliography: paper.bib
---

# Summary

<!--Follow similar spiel to OPTIMADE paper:-->
<!--- advent of high-throughput computing, software and theory maturity, availability of compute power have lead to explosion of computational data.-->
<!--- can be directly compared to high-quality measurements of crystal structures curated over many years-->
<!--- this data is increasingly being made available via public APIs, such as...-->
<!--- The OPTIMADE API specification was created to enable interoperability and machine-actionable APIs from multiple data providers-->

In recent decades, improvements in algorithms, hardware, and theory have enabled crystalline materials to be studied at the atomistic level with great accuracy and speed.

> To enable dissemination, reproducibility, and reuse, many digital crystal structure databases have been created and curated, ready for comparison with existing infrastructure storing structural characterizations of real crystals.

Author: ```suggestion
To enable dissemination, reproducibility, and reuse, many digital crystal structure databases have been created and curated, ready for comparison with existing infrastructure that stores structural characterizations (e.g., diffraction) of real crystals.
```

Reviewer: Thanks @ml-evs for the work you put already in there. 

I think there could be more emphases on "use-cases". You rightly dissect the functionality into routing, filter-language, db/resource-handling. I think we could find a few examples of how to recombine those functions in multiple abstract scenarios rather than "simply" stating what MP, MC, NOMAD are doing? I feel readers would identify more with more general use-cases than with our specifics?
- use OPT as a standalone app, e.g. to server your data
- integrate it into an existing app, e.g. on top of an existing database backend [e.g. what MP, MC, NOMAD are doing]
- use the filter independent of an app and routing (e.g. NOMAD uses it in our UI/API searches)
- maybe even more novel ideas like this "run a auto-configured OPT docker container" directly on your data

Reviewer: Could you remind me on the scope of the paper? I assume it addresses ML, data science people? Is this a physics or materials science specific community? What are you page boundaries?

Author: Hi @markus1978, sorry for missing this earlier.

> I think there could be more emphases on "use-cases". You rightly dissect the functionality into routing, filter-language, db/resource-handling. I think we could find a few examples of how to recombine those functions in multiple abstract scenarios rather than "simply" stating what MP, MC, NOMAD are doing? I feel readers would identify more with more general use-cases than with our specifics?

I think you are probably right, and I like your suggestions. Of course, these are things we said we would write anyway :)

Author: > Could you remind me on the scope of the paper? I assume it addresses ML, data science people? Is this a physics or materials science specific community? What are you page boundaries?

JOSS is multi-disciplinary, but purely focused on the open software. The review will be done a software developer with expertise in matsci (basically like us), and I find that they focus more on the repository than the contents of the paper itself (compared to a scientific journal). There are no page boundaries but the papers are typically short (1-4 pages) that acts as adverts for the code itself.

Consider the following query URL:

```
optimade.example.org/v1/structures?filter=chemical_formula_anonymous="ABC"
```

This query should match any crystal structures in the database with a composition that consists of any three elements in a 1:1:1 ratio. The "anatomy" of this query is displayed in Figure \ref{fig:query}.

1. After routing the query to the appropriate `/structures/` endpoint adhering to `v1` of the specification, the filter string `chemical_formula_anonymous="ABC"` is tokenized and parsed into an abstract tree by a `FilterParser` object using the Lark parsing library [@Lark] against the Extended Backus-Naur Form (EBNF) grammar defined by the specification.
2. The abstract tree is then transformed by a `FilterTransformer` object into a database query specific to the configured back-end for the server.
This transformation can include aliasing and custom transformations such that the underlying database format can be accommodates.
3. The results from the database query are then deserialized by `EntryResourceMapper` objects into the OPTIMADE-defined data models and then re-serialized into JSON before being served to the user over HTTP.

> ![Anatomy of an OPTIMADE query handled by the library.\label{fig:query}](./query.pdf)

Reviewer 2: the figure uses the `optimade.obdx.science` domain rather than the `optimade.example.org` domain in the text. Might want to re-use `optimade.example.org`?

Author: Yes, this needs to be updated. Happy to take any aesthetic suggestions too; personally I would rather either make it more whacky (colours, less boring boxes) or much more formal, like a UML figure or flowchart.

- NOMAD [@nomad] uses `optimade-python-tools` as a library to extend its existing web app with OPTIMADE API routes.
 It uses the Elasticsearch implementation to filter millions of structures from published first-principles calculations provided by users and other projects.
NOMAD also uses the filtering module in its own API to expose the OPTIMADE filter language in the user-centric web interface search bar.
NOMAD uses a released version of `optimade-python-tools` and all necessary customization can be realized via configuration and sub-classing.
- Materials Cloud [@MaterialsCloud] uses `optimade-python-tools` as a library to provide an OPTIMADE API entry to archived computational materials studies, created with the AiiDA [@AiiDA] Python framework and published through their archive.
In this case, each individual study and archive entry has its own database and separate API entry.
The Python classes within the `optimade` package have been extended to make use of AiiDA and its underlying [PostgreSQL](https://postgresql.org) storage engine.

> - The `optimade.adapters` module from the `optimade-python-tools` library is used in a graphical web client for OPTIMADE API implementations hosted on Materials Cloud [@MaterialsCloudClient].
> It allows users to search through OPTIMADE API implementations using widgets as well as typing in a filter string raw.
> The found structures can be inspected graphically and either converted and downloaded in formats provided by conversion functions in the `optimade.adapters` module or used in another Materials Cloud web tool, where the structure is automatically downloaded, converted, and transferred in the backgorund, partly using the `optimade.adapters` module.

Author: ```suggestion
- The `optimade.adapters` module from the `optimade-python-tools` library is used in a graphical web client hosted on Materials Cloud [@MaterialsCloudClient].
It allows users to query OPTIMADE API implementations using user-friendly widgets as well as raw filter strings.
The client uses the registry of known OPTIMADE providers to allow easy switching between databases.
The crystal structures returned can be inspected visually and either downloaded in formats provided by conversion functions in the `optimade.adapters` module, or used seamlessly within other Materials Cloud web tools, where the structure is automatically validated and transferred in the background, partly using the `optimade.adapters` module.
```

Thoughts @CasperWA?

Reviewer 3: Great re-write!

Should it be "raw OPTIMADE filter strings" maybe? Or is this implied well-enough?

Author: I think it is implied, I had "OPTIMADE" there originally but I think we say it enough times in the paper :sweat_smile:

Author: I will commit and rebuild the paper but feel free to change it again

  - name: Shyam Dwaraknath
    orcid: 0000-0003-0289-2607
    affiliation: 4
  - name: Markus Scheidgen
    orcid: 0000-0002-8038-2277
    affiliation: "5, 6"
  - name: Ádám Fekete
    orcid: 0000-0002-6263-897X
    affiliation: "1, 8, 9"
  - name: Donald Winston
    orcid: 0000-0002-8424-0604
    affiliation: "4, 7"
affiliations:
 - name: Institut de la Matière Condensée et des Nanosciences, Université catholique de Louvain, Chemin des Étoiles 8, Louvain-la-Neuve 1348, Belgium
   index: 1
 - name: Theory of Condensed Matter Group, Cavendish Laboratory, University of Cambridge, J. J. Thomson Avenue, Cambridge, CB3 0HE, U.K.
   index: 2
 - name: Theory and Simulation of Materials (THEOS), Faculté des Sciences et Techniques de l'Ingénieur, École Polytechnique Fédérale de Lausanne, CH-1015 Lausanne, Switzerland
   index: 3
 - name: Lawrence Berkeley National Laboratory, Berkeley, CA, USA
   index: 4
 - name: Fritz-Haber-Institut der Max-Planck-Gesellschaft, Faradayweg 4-6, 14195, Berlin, Germany
   index: 5
 - name: Humboldt-Universität zu Berlin, Institut für Physik and IRIS Adlershof, 12489 Berlin, Germany
   index: 6
 - name: Polyneme LLC, New York, NY, USA
   index: 7

>  - name: KCL

Reviewer 4: ```suggestion
 - name: Department of Physics, King's College London, Strand, London WC2R 2LS, United Kingdom
```

   index: 8

>  - name: Namur?

Reviewer 4: ```suggestion
 - name: Department of Physics and Namur Institute of Structured Materials, University of Namur, Rue de Bruxelles 51, 5000 Namur, Belgium
```
